lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/UTxOAssumptions.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Cardano.Wallet.Write.Tx.Balance.UTxOAssumptions where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( KeyHash, ScriptTemplate )
+import Cardano.Wallet.Address.Derivation.SharedKey
+    ( SharedKey )
+import Cardano.Wallet.Flavor
+    ( Excluding
+    , KeyFlavorS (ByronKeyS, IcarusKeyS, ShelleyKeyS)
+    , KeyOf
+    , WalletFlavorS
+    , keyOfWallet
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address )
+import Cardano.Wallet.TxWitnessTag
+    ( TxWitnessTag (..) )
+
+import qualified Cardano.Address.Script as CA
+
+-- | Assumptions about the UTxO which are needed for coin-selection.
+data UTxOAssumptions
+    = AllKeyPaymentCredentials
+    -- Assumes all 'UTxO' entries have addresses with the post-Shelley
+    -- key payment credentials.
+    | AllByronKeyPaymentCredentials
+    -- Assumes all 'UTxO' entries have addresses with the boostrap/byron
+    -- key payment credentials.
+    | AllScriptPaymentCredentialsFrom
+    -- Assumes all 'UTxO' entries have addresses with script
+    -- payment credentials, where the scripts are both derived
+    -- from the 'ScriptTemplate' and can be looked up using the given function.
+        !ScriptTemplate
+        !(Address -> CA.Script KeyHash)
+
+utxoAssumptionsForWallet
+    :: forall s. Excluding '[SharedKey] (KeyOf s)
+    => WalletFlavorS s -> UTxOAssumptions
+utxoAssumptionsForWallet walletFlavor =
+    case keyOfWallet walletFlavor of
+        ByronKeyS -> AllByronKeyPaymentCredentials
+        IcarusKeyS -> AllByronKeyPaymentCredentials
+        ShelleyKeyS -> AllKeyPaymentCredentials
```

Comment:
Putting `utxoAssumptionsForWallet` in the same module as `UTxOAssumptions` goes against our efforts of defining a boundary between the wallet and the balancing library. Here, `Cardano.Wallet.Write.Tx.Balance` is still dependent on `SharedKey` and now also `Wallet.Flavor`.

Additionally I think `UTxOAssumptions` belongs next to `selectAssets`.

I would therefore suggest to keep `UTxOAsssumptions` in `Balance.Tx` (as `selectAssets` is still there), and to put `utxoAssumptionsForWallet` in.... some other module... Exactly where? I'm not sure:

- I started adding `ChangeAddressGen` constructors in Wallet.hs https://github.com/input-output-hk/cardano-wallet/blob/master/lib/wallet/src/Cardano/Wallet.hs#L3660-L3682
- We have https://github.com/input-output-hk/cardano-wallet/blob/master/lib/wallet/src/Cardano/Wallet/Transaction/Built.hs#L1-L15
- There's both `Cardano.Wallet.Transaction` and `Cardano.Wallet.Shelley.Transaction` as you know